awscrt/eventstream.py

```diff
@@ -0,0 +1,197 @@
+            If successful, the Future will contain None.
+            Otherwise it will contain an exception.
+            If the connection is successful, it is accessible via `handler.connection`.
+        """
+
+        if not socket_options:
+            socket_options = SocketOptions()
+
+        future = Future()
+
+        # Connection is not made available to user until setup callback fires
+        connection = cls(host_name, port)
+
+        # We must be careful to avoid circular references that prevent the connection from getting GC'd.
+        # Only the internal _on_setup callback binds strong references to the connection and handler.
+        # This is ok because it fires exactly once, and references to it are cleared afterwards.
+        # All other callbacks must bind weak references to the handler,
+        # or references to futures within the connection rather than the connection itself.
+        handler_weakref = weakref.ref(handler)
+
+        connection._binding = _awscrt.event_stream_rpc_client_connection_connect(
+            host_name,
+            port,
+            bootstrap,
+            socket_options,
+            tls_connection_options,
+            partial(cls._on_connection_setup, future, handler, connection),
+            partial(cls._on_connection_shutdown, connection.shutdown_future, handler_weakref),
+            partial(cls._on_protocol_message, handler_weakref))
+
+        return future
+
+    @staticmethod
+    def _on_connection_setup(bound_future, bound_handler, bound_connection, error_code):
+        try:
+            if error_code:
+                e = awscrt.exceptions.from_code(error_code)
+                bound_future.set_exception(e)
+            else:
+                bound_handler.connection = bound_connection
+                bound_handler.on_connection_setup()
+                bound_future.set_result(None)
+        except Exception as e:
+            # user callback had unhandled exception, set future as failed
+            bound_future.set_exception(e)
+            raise
+
+    @staticmethod
+    def _on_connection_shutdown(bound_future, bound_weak_handler, error_code):
+        reason = awscrt.exceptions.from_code(error_code) if error_code else None
+        try:
+            handler = bound_weak_handler()
+            if handler:
+                handler.on_connection_shutdown(reason=reason)
+        finally:
+            # user callback had unhandled exception, use finally to ensure future gets set
+            if reason:
+                bound_future.set_exception(reason)
+            else:
+                bound_future.set_result(None)
+
+    @staticmethod
+    def _on_protocol_message(bound_weak_header, headers, payload, message_type, flags):
+        handler = bound_weak_handler()
+        if handler:
+            handler.on_protocol_message(
+                headers=headers,
+                payload=payload,
+                message_type=message_type,
+                flags=flags)
```

Comment:
Headers, payload, message type, and flag types to be understood later? All are probably built-in types except Header which needs to be bound for creation and correct interpretation in Python.

Reply:
yeah, this will be in next PR